server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java

```diff
@@ -29,30 +31,49 @@ public static void preProcess(SearchContext context) {
         if (context.aggregations() == null) {
             return;
         }
-        BucketCollector bucketCollector;
-        try {
-            context.aggregations().aggregators(context.aggregations().factories().createTopLevelAggregators());
-            bucketCollector = MultiBucketCollector.wrap(true, List.of(context.aggregations().aggregators()));
-            bucketCollector.preCollection();
-        } catch (IOException e) {
-            throw new AggregationInitializationException("Could not initialize aggregators", e);
-        }
-        final Collector collector;
-        if (context.aggregations().factories().context() != null
-            && context.aggregations().factories().context().isInSortOrderExecutionRequired()) {
-            TimeSeriesIndexSearcher searcher = new TimeSeriesIndexSearcher(context.searcher(), getCancellationChecks(context));
-            searcher.setMinimumScore(context.minimumScore());
-            searcher.setProfiler(context);
+        final Supplier<BucketCollector> bucketCollectorSupplier = () -> {
             try {
-                searcher.search(context.rewrittenQuery(), bucketCollector);
+                Aggregator[] aggregators = context.aggregations().factories().createTopLevelAggregators();
+                context.aggregations().aggregators(aggregators);
+                BucketCollector bucketCollector = MultiBucketCollector.wrap(true, List.of(aggregators));
+                bucketCollector.preCollection();
+                return bucketCollector;
             } catch (IOException e) {
-                throw new AggregationExecutionException("Could not perform time series aggregation", e);
+                throw new AggregationInitializationException("Could not initialize aggregators", e);
             }
-            collector = BucketCollector.NO_OP_COLLECTOR;
+        };
+        final Supplier<Collector> collectorSupplier;
+        if (context.aggregations().isInSortOrderExecutionRequired()) {
+            executeInSortOrder(context, bucketCollectorSupplier.get());
+            collectorSupplier = () -> BucketCollector.NO_OP_COLLECTOR;
         } else {
-            collector = bucketCollector.asCollector();
+            collectorSupplier = () -> bucketCollectorSupplier.get().asCollector();
+        }
+        context.aggregations().registerAggsCollectorManager(new CollectorManager<>() {
+            @Override
+            public Collector newCollector() {
+                return collectorSupplier.get();
```

Comment:
would it be possible to include the logic in the two suppliers in newCollector directly, and avoid requiring the two suppliers?

Reply:
No without having to duplicate the buckets collector method. This object is needed for running tsdb aggregations and that's the reason I am using two suppliers.

Reply:
right I missed that.

Reply:
Would something like the following work?

```
diff --git a/server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java b/server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java
--- a/server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java	(revision be8a385062a3aa97879cd5abea156266e80e54d9)
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java	(date 1690221857262)
@@ -31,28 +31,16 @@
         if (context.aggregations() == null) {
             return;
         }
-        final Supplier<BucketCollector> bucketCollectorSupplier = () -> {
-            try {
-                Aggregator[] aggregators = context.aggregations().factories().createTopLevelAggregators();
-                context.aggregations().aggregators(aggregators);
-                BucketCollector bucketCollector = MultiBucketCollector.wrap(true, List.of(aggregators));
-                bucketCollector.preCollection();
-                return bucketCollector;
-            } catch (IOException e) {
-                throw new AggregationInitializationException("Could not initialize aggregators", e);
-            }
-        };
-        final Supplier<Collector> collectorSupplier;
         if (context.aggregations().isInSortOrderExecutionRequired()) {
-            executeInSortOrder(context, bucketCollectorSupplier.get());
-            collectorSupplier = () -> BucketCollector.NO_OP_COLLECTOR;
-        } else {
-            collectorSupplier = () -> bucketCollectorSupplier.get().asCollector();
+            executeInSortOrder(context, newBucketCollector(context));
         }
         context.aggregations().registerAggsCollectorManager(new CollectorManager<>() {
             @Override
             public Collector newCollector() {
-                return collectorSupplier.get();
+                if (context.aggregations().isInSortOrderExecutionRequired()) {
+                    return BucketCollector.NO_OP_COLLECTOR;
+                }
+                return newBucketCollector(context).asCollector();
             }
 
             @Override
@@ -65,6 +53,18 @@
         });
     }
 
+    private static BucketCollector newBucketCollector(SearchContext context) {
+        try {
+            Aggregator[] aggregators = context.aggregations().factories().createTopLevelAggregators();
+            context.aggregations().aggregators(aggregators);
+            BucketCollector bucketCollector = MultiBucketCollector.wrap(true, List.of(aggregators));
+            bucketCollector.preCollection();
+            return bucketCollector;
+        } catch (IOException e) {
+            throw new AggregationInitializationException("Could not initialize aggregators", e);
+        }
+    }
+
     private static void executeInSortOrder(SearchContext context, BucketCollector collector) {
         TimeSeriesIndexSearcher searcher = new TimeSeriesIndexSearcher(context.searcher(), getCancellationChecks(context));
         searcher.setMinimumScore(context.minimumScore());
```

I find it odd that we create a supplier that we immediately call get on. The only place where we do need to create a new instance at each call is the newCollector method but we can call a method for that. I find that the resulting code is easier to follow.

Reply:
I took half of your suggestion by introducing the `newBucketCollector` method. Still use a supplier on the `newCollector` method, i prefer not to perform the check in each call.